### PR TITLE
Refactor: Place matching use strategy pattern

### PR DESCRIPTION
**Pull request by timwangmusic**
## Description
Clearly describe changes in your PR, new feature or bug fixes.

## Solution
* Describe your code changes in detail for reviewers as well as documentation purposes. 
* For bug fixes, explain the solutions and how they fix the issues.
* For new features, describe the high-level ideas.

## Testing
- [x] Integration testing on Heroku staging
- [x] Added new unit tests

## Checks
- [x] Have you removed commented code?
- [x] Have you used gofmt to format your code?

### CODE REVIEW

File: planner/solver.go [hunk 1]
```diff
@@ -46,12 +46,13 @@ func (ps PlanningSolution) Key() float64 {
...
 
 type Solver struct {
-	Searcher                *iowrappers.PoiSearcher
-	UserRatingsCountMatcher matching.Matcher
-	TimeMatcher             matching.Matcher
-	PriceRangeMatcher       matching.Matcher
-	placeDedupeCountLimit   int
-	nearbyCitiesCountLimit  int
+	Searcher               *iowrappers.PoiSearcher
+	placeMatcher           *PlaceMatcher
```
[the remark this answers is not in the document]
Comment (rwangsc18) [reply planner/solver.go:50]: Then later we can just run all these filtering operations in a for loop?
```
for _, matcher : range matchers {
  result := matcher.match(req, result)
}
```

**Review by rwangsc18 — COMMENTED** [review]
LGTM. I wonder if multiple matchers in Solver can be combined into a slice of matcher interface pointers.

File: planner/solver.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (timwangmusic) [reply planner/solver.go:50]: But how do we determine which matcher type to add at runtime?

[the remark this answers is not in the document]
Comment (rwangsc18) [reply planner/solver.go:50]: I think we can just add all required matchers (time, price, userRating) in INIT() of solver as a slice of interface type maching.Matcher. I am not super familiar with golang interface, but in C++ there is virtual base class and pointers to base class with derived class objects, so we can use runtime polymorphism.

[the remark this answers is not in the document]
Comment (rwangsc18) [reply planner/solver.go:50]: I have searched a bit about golang interface. There is some example close to what I mentioned. https://medium.com/@sonasingh46/strategy-pattern-in-golang-cc8747d4dc25. Feel free to take it as reference or not.

[the remark this answers is not in the document]
Comment (timwangmusic) [reply planner/solver.go:50]: I see what you mean now. However we will lose some logging capabilities, maybe we can add them in the filters themselves.

[the remark this answers is not in the document]
Comment (timwangmusic) [reply planner/solver.go:50]: I tried to add those matchers in a concrete matchers field, it is feasible to do what you described in a loop. However it seems to complicate things and we lose the logging granularity. Let me know what you think.

[the remark this answers is not in the document]
Comment (rwangsc18) [reply planner/solver.go:50]: It probably does take some more work to refactor things, but it should be okay to leave things as they are, if we don't expect to add many new matchers in the near future. I am okay with submitting this PR.

**Review by rwangsc18 — APPROVED** [review]

File: planner/solver.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (rwangsc18) [reply planner/solver.go:50]: Regarding with the logging, we can probably add a data member "string name" in the matcher interface. When adding actual matchers in the INIT, we just need to give them the right names, and we can also log them with proper type info.

**Comment by sananand007** [comment]
Can you please add a description of what we are solving with this PR ?

File: planner/solver.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (timwangmusic) [reply planner/solver.go:50]: Added `MatcherName` to the `Matcher` interface, now we can get appropriate logging details.

**Review by rwangsc18 — APPROVED** [review]
Thanks! LGTM.

**Pull request merged**